### ISSUE TITLE
Modify Python Logging

### ISSUE THREAD

**Issue by rldotai** [apache/incubator-mxnet#1960]
TL;DR: At present, the way logging is done causes problems with newer versions of the Jupyter/IPython notebook, among other things.

When I installed MXNet on a new machine recently and ran some notebooks to test it out I was somewhat confused because none of the callbacks for reporting training speed/accuracy were printed as output. 

For example, the [tutorial notebook has messages printed like](https://github.com/dmlc/mxnet/blob/master/example/notebooks/tutorial.ipynb):

```
INFO:root:Start training with [gpu(0)]
INFO:root:Epoch[0] Batch [200]  Speed: 70941.64 samples/sec Train-accuracy=0.389050
INFO:root:Epoch[0] Batch [400]  Speed: 97857.94 samples/sec Train-accuracy=0.646450
INFO:root:Epoch[0] Batch [600]  Speed: 70507.97 samples/sec Train-accuracy=0.743333
INFO:root:Epoch[0] Resetting Data Iterator
INFO:root:Epoch[0] Train-accuracy=0.743333
INFO:root:Epoch[0] Time cost=1.069
INFO:root:Epoch[0] Validation-accuracy=0.950800
...
```

However, these weren't showing up for me (although I recalled seeing them on my other machine).

This behavior turns out to be due to changes made to Jupyter's logging, where messages to the root logger (which is what gets called when invoking `logging.info`, `logging.debug`, etc.) no longer print to `stdout`, or maybe `stderr` doesn't show up in the notebook anymore.
[See the relevant Python docs](https://docs.python.org/3/library/logging.html#logging.debug)

This can be fixed reasonably quickly via adding the following code at the top of the notebook:

``` python
import logging
import sys
root_logger = logging.getLogger()
stdout_handler = logging.StreamHandler(sys.stdout)
root_logger.addHandler(stdout_handler)
root_logger.setLevel(logging.DEBUG)
```

But it's somewhat inelegant, since the root logger may be used by other modules whereas we presumably just want the log output from MXNet and the callbacks we are using.
It also causes issues in the documentation, where the [generated documents refer to objects by the full Python path](http://mxnet.readthedocs.org/en/latest/packages/python/module.html#mxnet.module.base_module.BaseModule), e.g.,

```
class mxnet.module.base_module.BaseModule(logger=<module 'logging' from '/usr/lib/python2.7/logging/__init__.pyc'>)
```

I am not sure about what the ideal fix ought to be, but maybe it would be better to define a specific logger for MXNet rather than just using the methods of the `logging` module directly.
If you guys want, I can put together a pull request at some point next week when I'm less busy.

**Comment by piiswrong** [comment]
Usually you put logging.basicConfig(level=logging.DEBUG) at the start of every program to see the log.

It's easy to forget this though.

@tqchen @antinucleon Maybe we should put logging.basicConfig(level=logging.DEBUG) into mxnet's init?

**Comment by pluskid** [comment]
Maybe put all the logger default value of all methods to be `None`, and then have a internal `mxnet.logging` module defining some method like

```python
def get_logger(logger):
    if logger is None:
        # construct or re-use some global default logger
    else:
        return logger
```

**Comment by rldotai** [comment]
You might not even have to do that much.
With [`logging.getLogger(<name>)`](https://docs.python.org/3/library/logging.html#logging.getLogger), for the same value of `name` you're guaranteed to get the same logger object.

So in modules where you want to use logging, you could just have

```python
...
import logging
...
# A module scoped reference to the logger named "MXNet"
_logger = logging.getLogger('MXNet')

def foo(x, logger=None):
    if logger is None:
        logger = _logger
    ...

# Or alternatively, you could call `getLogger()`
class Bar:
    def __init__(self, logger=None, *args, **kwargs):
        if logger is None:
            self.logger = logging.getLogger('MXNet')
        else:
            self.logger = logger
```

The logger named "MXNet" could be configured, as you suggested, in the package's `__init__.py`.

I've just been defining a separate logger object and then passing that as a keyword argument when appropriate, and so far it seems to work well.  
I am pretty sure that this would be easy to set up, but I'd want to spend more time looking at the code to make sure I'm not inadvertently breaking something.

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
